sys/arduino/include/SPI.h

```diff
@@ -0,0 +1,163 @@
+}
+
+/**
+ * @name    Arduino compatible SPI modes
+ * @{
+ */
+#define SPI_MODE0       (0)     /**< CPOL=0, CPHA=0 */
+#define SPI_MODE1       (1)     /**< CPOL=0, CPHA=1 */
+#define SPI_MODE2       (2)     /**< CPOL=1, CPHA=0 */
+#define SPI_MODE3       (3)     /**< CPOL=1, CPHA=1 */
+/** @} */
+
+/**
+ * @name    Arduino compatible bit order values for SPI
+ * @{
+ */
+#define MSBFIRST        (0)     /**< most significat bit first */
+/** @} */
+
+/**
+ * @brief   Arduino SPI configuration interface
+ */
+class SPISettings {
+    public:
+    /**
+     * @brief   Create a new SPI settings instance
+     *
+     * @param   clock       SPI clock in Hz to use
+     * @param   bitOrder    Has to be `MSBFIST`, for compatibility only
+     * @param   dataMode    SPI mode to use
+     */
+    SPISettings(uint32_t clock, uint8_t bitOrder, uint8_t dataMode);
+
+    /**
+     * @brief   Create a new SPI settings instance with default settings
+     */
+    SPISettings() : SPISettings(4000000, MSBFIRST, SPI_MODE0) { }
+
+    private:
+    spi_mode_t mode;
+    spi_clk_t clock;
+
+    friend class SPIClass;
+};
+
+/**
+ * @brief   Arduino SPI interface
+ *
+ * @warning Wrap all SPI transfers in `SPI.beginTransaction()` and `SPI.endTransaction()`
+ *
+ * The official Arduino SPI-API allows to use SPI transfers without having to
+ * call `SPI.beginTransaction()`, but discourages to do so. The RIOT API does
+ * not provide this feature, instead a call to `SPI.beginTransaction()` is
+ * mandatory. However, most Arduino code already does this and the remaining
+ * code should be fixed anyway.
+ */
+class SPIClass {
+    public:
+    /**
+     * @brief   Create a new SPI interface instance
+     * @param   spi_dev     The RIOT SPI device to use under the hood
+     */
+    SPIClass(spi_t spi_dev);
+
+    /**
+     * @brief   Create a new SPI interface instance for SPI device 0
+     * @param   uc_pinMISO  Ignored, for compatibility only
+     * @param   uc_pinSCK   Ignored, for compatibility only
+     * @param   uc_pinMOSI  Ignored, for compatibility only
+     * @param   uc_pinSS    Ignored, for compatibility only
+     * @param   uc_mux      Ignored, for compatibility only
+     */
+    SPIClass(uint8_t uc_pinMISO, uint8_t uc_pinSCK, uint8_t uc_pinMOSI,
+             uint8_t uc_pinSS, uint8_t uc_mux) : SPIClass(SPI_DEV(0))
+    {
+        (void)uc_pinMISO;
+        (void)uc_pinSCK;
+        (void)uc_pinMOSI;
+        (void)uc_pinSS;
+        (void)uc_mux;
+    }
+
+    /**
+     * @brief   Transfer a single byte of data
+     * @param[in]   data        Byte to send
+     * @return  The received byte
+     */
+    uint8_t transfer(uint8_t data) {
+        transfer(&data, sizeof(data));
+        return data;
+    }
+
+    /**
+     * @brief   Transfer two bytes of data
+     * @param[in]       data        The two bytes to send
+     * @return  The received two bytes
+     */
+    uint16_t transfer16(uint16_t data) {
+        transfer(&data, sizeof(data));
```

Comment:
This has a potential endiness problem. In this implementation you're using the native endianess of the selected CPU. In most cases that is Little Endian.

To be honest I don't know enough about SPI and its bit order. Is that bit order or byte order? Arduino's implementation does something with the byte ordering. And it seems to think that with MSBFIRST, that the MS byte comes first. Now, combine that with the fact that we currently `assert(bitOrder == MSBFIRST);` and we for sure have a problem on Little Endian machines.

Reply:
I'll check the Arduino implementation on the byteorder here. Thanks for pointing that out.

`bitOrder` is about which bit is first send over the wire. I think today it safe to assune msb first. But we can add lsb first later, if a use case still popa up

Reply:
Arduino is sending the most significant byte first, if the most significant bit is first send over the wire: https://github.com/arduino/ArduinoCore-megaavr/blob/7df9f211927b902a67ed322734a32ebfe89dd34b/libraries/SPI/src/SPI.cpp#L245-L254

I do not understand why this is tied together, as these are completely unrelated things... But anyhow, I will just follow suit for compatibility.